### ISSUE TITLE
New representation for scope handler tests

### ISSUE THREAD

**Issue by pokey** [cursorless-dev/cursorless#1524]
# The problem

We need a better way to represent our test cases for scope handlers, because today we end up needing to record a bunch of commands to test edge cases, and the tests aren't easy to read.  In addition, once #1523 is merged, we expect that most scope type development will be done by viewing the output of the scope visualizer, so we'd like test cases that mirror the scope visualizer as closely as possible

# The solution

We'd like an ascii representation that is readable by humans and parsable by machine.  The idea is as follows:

- We intersperse comment lines among the source code that point out regions in the line above
- The comment lines begin with the language's comment token (eg `//`), followed by `!`
- Each source code line is indented by extra whitespace at the beginning to allow the comment lines to be able to point at the start of the line above if necessary.  If we didn't do this, then a source code line could begin at  the same spot as the `//` on the line below, so we couldn't point out the beginning of the line.
- The content range is marked by `{^^^}`
- The domain is marked by `[---]`
- The removal range is marked by `(xxx)`
- The iteration scope is marked by `<***>`
- We need to figure out `interior` / `leadingDelimiter` / `trailingDelimiter`
  - Would like to avoid unicode if possible
  - Here are suggestions from chatGPT:
    - The `interior` is marked by `/###\`
    - The `leadingDelimiter` is marked by `«@@@»`
    - The `trailingDelimiter` is marked by `§$$$¶`
- If any range is equal to `contentRange`, it does not get a line, as that is the implicit assumption
- In each of the above, the surrounding brackets are _outside_ of the range, so that we can handle empty ranges by using eg `{}`
- Each range type (eg content, domain, removal, etc) gets its own comment line
- If there are overlapping ranges, the range with later start or earlier end gets its own set of lines after the other scope, beginning with `//!2` (or `//!3`, etc).  Note that ranges are overlapping even if they are 1 character apart, because otherwise there's no room for the brackets
- The start of the file will contain metadata lines to indicate scope type being tested and language id, eg
  ```ts
   //!! scope: {type: namedFunction}
   //!! languageId: typescript, plaintext
   ```
- Note that multiple examples of the same scope type could be present in one file, so that we don't need a million test files like we have today
- Note that we don't handle indicating `isPreferredOver`; need special tests for that

## Components

- [ ] Automatically generate test case from source code.  Can probably steal some code from #1523 
- [ ] Parse output to run as a test.  Before running the scope handler, it should remove all the special comment lines and dedent every line by the correct amount.  Note that it should be able to only look at the delimiters (eg `{}`), as the internal markers (eg `^^^`) are just for readability

## Considerations

- [ ] #2064

### `value`

```ts
//!! scopeType: {type: value}
//!! languageId: typescript

     const removalRange = getRelatedRange(match, scopeTypeType, "removal", true);
//!                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ 
//!                    xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx 
//!  ----------------------------------------------------------------------------
```

### `name`
```ts
//!! scopeType: {type: name}
//!! languageId: typescript

     const removalRange = getRelatedRange(match, scopeTypeType, "removal", true);
//!        ^^^^^^^^^^^^                                                          
//!  ----------------------------------------------------------------------------
```

### `statement`
```ts
//!! scopeType: {type: statement}
//!! languageId: typescript

     const removalRange = getRelatedRange(match, scopeTypeType, "removal", true);
//!  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

### `namedFunction`
```ts
//!! scopeType: {type: namedFunction}
//!! languageId: typescript

      function aaa() {
//!1 {^^^^^^^^^^^^^^^^
        function bbb() {
//!2   {^^^^^^^^^^^^^^^^
          const ccc = "ddd";
        }
//!2  ^^^}
      }
//!1  ^}
```

### `collectionKey`
```ts
//!! scopeType: {type: collectionKey}
//!! languageId: typescript

     const aaa = {
       bbb: {
//!1   ^^^   
//!1  [------
         ccc: "ddd",
//!2     ^^^        
//!2    [-----------]
       },
//!1   --]
     }
```

### `token`
```
#!! scopeType: {type: token}
#!! languageId: plaintext

     foo.bar.baz
#!1  ^^^        
#!2     ^       
#!3      ^^^    
#!4         ^   
#!5          ^^^
```

Here are some alternative representations to consider:

<details>
<summary>Merge ranges onto one line</summary>

### `value`

```ts
//!! scopeType: {type: value}
//!! languageId: typescript

     const removalRange = getRelatedRange(match, scopeTypeType, "removal", true);
//!  ------------------xxx^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^-
```

### `name`
```ts
//!! scopeType: {type: name}
//!! languageId: typescript

     const removalRange = getRelatedRange(match, scopeTypeType, "removal", true);
//!  ------^^^^^^^^^^^^----------------------------------------------------------
```

### `statement`
```ts
//!! scopeType: {type: statement}
//!! languageId: typescript

     const removalRange = getRelatedRange(match, scopeTypeType, "removal", true);
//!  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

### `namedFunction`
```ts
//!! scopeType: {type: namedFunction}
//!! languageId: typescript

      function aaa() {
//!1 {^^^^^^^^^^^^^^^^
        function bbb() {
//!2   {^^^^^^^^^^^^^^^^
          const ccc = "ddd";
        }
//!2  ^^^}
      }
//!1  ^}
```

### `collectionKey`
```ts
//!! scopeType: {type: collectionKey}
//!! languageId: typescript

     const aaa = {
       bbb: {
//!1   ^^^ 
//!1  [------
         ccc: "ddd",
//!2     ^^^--------
       },
//!1   --]
     }
```

### `token`
```
#!! scopeType: {type: token}
#!! languageId: plaintext

     foo.bar.baz
#!1  ^^^        
#!2     ^       
#!3      ^^^    
#!4         ^   
#!5          ^^^
```

</details>

<details>
<summary>Always surrounding brackets</summary>

### `value`

```ts
//!! scopeType: {type: value}
//!! languageId: typescript

     const removalRange = getRelatedRange(match, scopeTypeType, "removal", true);
//!                      {^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^}
//!                   (xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx)
//! [----------------------------------------------------------------------------]
```

### `name`
```ts
//!! scopeType: {type: name}
//!! languageId: typescript

     const removalRange = getRelatedRange(match, scopeTypeType, "removal", true);
//!       {^^^^^^^^^^^^}
//! [----------------------------------------------------------------------------]
```

### `statement`
```ts
//!! scopeType: {type: statement}
//!! languageId: typescript

     const removalRange = getRelatedRange(match, scopeTypeType, "removal", true);
//! {^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^}
```

### `namedFunction`
```ts
//!! scopeType: {type: namedFunction}
//!! languageId: typescript

      function aaa() {
//!  {^^^^^^^^^^^^^^^^
        function bbb() {
//!2   {^^^^^^^^^^^^^^^^
          const ccc = "ddd";
        }
//!2  ^^^}
      }
//!   ^}
```

### `collectionKey`
```ts
//!! scopeType: {type: collectionKey}
//!! languageId: typescript

     const aaa = {
       bbb: {
//!   {^^^}
//!   [------
         ccc: "ddd",
//!2    {^^^}
//!2    [-----------]
       },
//!    --]
     }
```

### `token`
```
#!! scopeType: {type: token}
#!! languageId: plaintext

     foo.bar.baz
#!  {^^^}  {^}
#!2    {^}  {^^^}
#!3     {^^^}
```

</details>

<details>
<summary>Internal markers on every line of range</summary>

### `value`, `name`, `statement`, and `token`

(as above)

### `namedFunction`
```ts
//!! scopeType: {type: namedFunction}
//!! languageId: typescript

      function aaa() {
//!  {^^^^^^^^^^^^^^^^
        function bbb() {
//!   ^^^^^^^^^^^^^^^^^^
//!2   {^^^^^^^^^^^^^^^^
          const ccc = "ddd";
//!   ^^^^^^^^^^^^^^^^^^^^^^
//!2  ^^^^^^^^^^^^^^^^^^^^^^
        }
//!   ^^^
//!2  ^^^}
      }
//!   ^}
```

### `collectionKey`
```ts
//!! scopeType: {type: collectionKey}
//!! languageId: typescript

     const aaa = {
       bbb: {
//!   {^^^}
//!   [------
         ccc: "ddd",
//!    -------------
//!2    {^^^}
//!2    [-----------]
       },
//!    --]
     }
```

</details>

<details>
<summary>No internal markers</summary>

### `value`

```ts
//!! scopeType: {type: value}
//!! languageId: typescript

     const removalRange = getRelatedRange(match, scopeTypeType, "removal", true);
//!                      {                                                      }
//! [                                                                            ]
//!                   (                                                         )
```

### `name`
```ts
//!! scopeType: {type: name}
//!! languageId: typescript

     const removalRange = getRelatedRange(match, scopeTypeType, "removal", true);
//!       {            }
//! [                                                                            ]
```

### `statement`
```ts
//!! scopeType: {type: statement}
//!! languageId: typescript

     const removalRange = getRelatedRange(match, scopeTypeType, "removal", true);
//! {                                                                            }
```

### `namedFunction`
```ts
//!! scopeType: {type: namedFunction}
//!! languageId: typescript

      function aaa() {
//!  {
        function bbb() {
//!2   {
          const ccc = "ddd";
        }
//!2     }
      }
//!    }
```

### `collectionKey`
```ts
//!! scopeType: {type: collectionKey}
//!! languageId: typescript

     const aaa = {
       bbb: {
//!   {   }
//!   [
         ccc: "ddd",
//!2    {   }
//!2    [           ]
       },
//!      ]
     }
```

### `token`
```
#!! scopeType: {type: token}
#!! languageId: plaintext

     foo.bar.baz
#!  {   }  { }
#!2    { }  {   }
#!3     {   }
```

</details>

<details>
<summary>Using `.` to indicate whitespace</summary>

### `value`

```ts
//!! scopeType: {type: value}
//!! languageId: typescript

     const removalRange = getRelatedRange(match, scopeTypeType, "removal", true);
//!  ....................{......................................................}
//! [............................................................................]
//!  .................(.........................................................)
```

### `name`
```ts
//!! scopeType: {type: name}
//!! languageId: typescript

     const removalRange = getRelatedRange(match, scopeTypeType, "removal", true);
//!  .....{............}.........................................................
//! [............................................................................]
```

### `statement`
```ts
//!! scopeType: {type: statement}
//!! languageId: typescript

     const removalRange = getRelatedRange(match, scopeTypeType, "removal", true);
//! {............................................................................}
```

### `namedFunction`
```ts
//!! scopeType: {type: namedFunction}
//!! languageId: typescript

      function aaa() {
//!  {................
        function bbb() {
//!2  .{................
          const ccc = "ddd";
        }
//!2  ...}
      }
//!   .}
```

### `collectionKey`
```ts
//!! scopeType: {type: collectionKey}
//!! languageId: typescript

     const aaa = {
       bbb: {
//!  .{...}..
//!  .[......
         ccc: "ddd",
//!2 ...{...}.......
//!2 ...[...........]
       },
//!  ....]
     }
```

### `token`
```
#!! scopeType: {type: token}
#!! languageId: plaintext

     foo.bar.baz
#!  {...}..{.}..
#!2  ..{.}..{...}
#!3  ...{...}...
```

</details>

<details>
<summary>Using `.` only where necessary</summary>

### `value`

```ts
//!! scopeType: {type: value}
//!! languageId: typescript

     const removalRange = getRelatedRange(match, scopeTypeType, "removal", true);
//!  .                 . {.                                                    .}
//! [.                 .                                                       ..]
//!                   (.                                                       .)
```

### `name`
```ts
//!! scopeType: {type: name}
//!! languageId: typescript

     const removalRange = getRelatedRange(match, scopeTypeType, "removal", true);
//!  .    {.          .}                                                        .
//! [.                                                                          .]
```

### `statement`
```ts
//!! scopeType: {type: statement}
//!! languageId: typescript

     const removalRange = getRelatedRange(match, scopeTypeType, "removal", true);
//! {.                                                                          .}
```

### `namedFunction`
```ts
//!! scopeType: {type: namedFunction}
//!! languageId: typescript

      function aaa() {
//!  {.
        function bbb() {
//!2   {.
          const ccc = "ddd";
        }
//!2    .}
      }
//!   .}
```

### `collectionKey`
```ts
//!! scopeType: {type: collectionKey}
//!! languageId: typescript

     const aaa = {
       bbb: {
//!   {. .}
//!   [.
         ccc: "ddd",
//!2    {. .}      .
//!2    [.         .]
       },
//!     .]
     }
```

### `token`
```
#!! scopeType: {type: token}
#!! languageId: plaintext

     foo.bar.baz
#!  {. .}. {.}
#!2    {.} .{. .}
#!3     {. .}
```

</details>

<details>
<summary>Using markers only where necessary</summary>

### `value`

```ts
//!! scopeType: {type: value}
//!! languageId: typescript

     const removalRange = getRelatedRange(match, scopeTypeType, "removal", true);
//!  .                 . {^                                                    ^}
//! [-                 .                                                       .-]
//!                   (x                                                       x)
```

### `name`
```ts
//!! scopeType: {type: name}
//!! languageId: typescript

     const removalRange = getRelatedRange(match, scopeTypeType, "removal", true);
//!  .    {^          ^}                                                        .
//! [-                                                                          -]
```

### `statement`
```ts
//!! scopeType: {type: statement}
//!! languageId: typescript

     const removalRange = getRelatedRange(match, scopeTypeType, "removal", true);
//! {^                                                                          ^}
```

### `namedFunction`
```ts
//!! scopeType: {type: namedFunction}
//!! languageId: typescript

      function aaa() {
//!  {^
        function bbb() {
//!2   {^
          const ccc = "ddd";
        }
//!2    ^}
      }
//!   ^}
```

### `collectionKey`
```ts
//!! scopeType: {type: collectionKey}
//!! languageId: typescript

     const aaa = {
       bbb: {
//!   {^ ^}
//!   [-
         ccc: "ddd",
//!2    {^ ^}      .
//!2    [-         -]
       },
//!     -]
     }
```

### `token`
```
#!! scopeType: {type: token}
#!! languageId: plaintext

     foo.bar.baz
#!  {^ ^}. {^} .
#!2    {^} .{^ ^}
#!3     {^ ^}
```

</details>

**Comment by pokey** [comment]
A few questions:

- should we put leading / trailing delimiter on same line as content range? They never overlap
- should we show removal range if it's just equal to `trailing ?? leading`?
- should we show leading / trailing / removal when it's just token delimiter ranges (ie whitespace)? On the one hand we could do that to write nice test cases for token removal, but on the other hand I worry it will lead to lots of noise for scopes that just get token removal by default, esp if removing them doesn't make much semantic sense. But for those where we actually want those token removal semantics it might be nice to have that tested 🤔. Maybe syntactic scopes could opt in / out of token removal behaviour?

cc/ @AndreasArvidsson @phillco

**Comment by AndreasArvidsson** [comment]
1. Please make an example of that I think it sounds reasonable
2. I believe so. It's not always obvious if it's going to be trailing or leading for a particular case and we might change that algorithm in the future.
3. I guess we could start with them and if it becomes too noisy we could remove them in the future?

**Comment by pokey** [comment]
> 1. Please make an example of that I think it sounds reasonable

```ts
//!! scopeType: {type: value}
//!! languageId: typescript

     const removalRange = getRelatedRange(match, scopeTypeType, "removal", true);
//!                    xxx^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ 
//!  ----------------------------------------------------------------------------
```

> 2. I believe so. It's not always obvious if it's going to be trailing or leading for a particular case and we might change that algorithm in the future.

Mixed feelings.  It will be extra noise I think, and while I do agree it can be unintuitive when you're just using Cursorless, in this case the leading / trailing are marked quite clearly visually.  But easiest to just show them for now, and then see how annoying it is.  If it feels like noise, I think just having a check if `target.removalRange.isRangeEqual(target.contentRange.union(target.leadingDelimiter ?? target.trailingDelimiter ?? target.contentRange)` is easy to do, and should be fairly intuitive

> 3. I guess we could start with them and if it becomes too noisy we could remove them in the future?

Sounds good

**Comment by pokey** [comment]
Yeah fair. I'm happy with that

**Comment by josharian** [comment]
Any thoughts about what to do about tabs that are part of ranges? There's no right answer about how many dashes/whatever to use under them. (Can discuss over voice if the answer is long.)

**Comment by pokey** [comment]
Yes tabs continue to be the bane of our existence in cursorless. See eg

- https://github.com/cursorless-dev/cursorless/issues/1586
- https://github.com/cursorless-dev/cursorless/issues/1741

I have no idea how to handle them tbh 😅. For some situations we could possibly read the tab width setting, as suggested here, but I don't think that helps us in this situation, as different users might have different settings, and we're baking in one setting to the recorded test.

My instinct is to avoid tabs wherever possible in our tests, but that will potentially cause skew from real usage in the case of Go, where I'm guessing most users will be using tabs due to gofmt

Very much open to ideas here

**Comment by josharian** [comment]
The only "works always" fix that I am aware of is to fight tabs with tabs: allows tabs anywhere in ranges as a no-op, and put them below any real tabs so that the alignment is always preserved. But that creates all sorts of knock-on problems.

Let's kick around a few ideas over voice sometime.

**Comment by josharian** [comment]
lolsob https://stackoverflow.com/questions/68333368/tab-stops-are-messed-up-after-rarer-unicode-characters

**Comment by josharian** [comment]
terrible idea: replace tabs with:

```
U+2409 ␉ SYMBOL FOR HORIZONTAL TABULATION
```

**Comment by josharian** [comment]
not that it helps us, but elastic tab stops are nice:

https://nick-gravgaard.com/elastic-tabstops/
https://pkg.go.dev/text/tabwriter

**Comment by josharian** [comment]
Another annoying question. (Sorry, thinking through all this stuff a bit.) How will this representation represent comment scopes if/when they exist?

**Comment by josharian** [comment]
FWIW, I think the "fixed but inconvenient length" problem (how do you keep symbol-by-symbol alignment in the face of grapheme clusters) may be easier: use the correct number of zero-width spaces to preserve alignment.

And yes, there's a opposite type of this category of problem, which is zero-width spaces in content. Maybe we can assume nobody uses those. 😅

**Comment by pokey** [comment]
> Another annoying question. (Sorry, thinking through all this stuff a bit.) How will this representation represent comment scopes? The scope system will pick up the annotations as comments themselves...

That's what the `!` is for in `//!`

**Comment by AndreasArvidsson** [comment]
I have an idea for a really simple format to test scopes
* Folder name denotes language. If all tests in this folders should be the same language there's really no point duplicating the language as a field in the fixture.
* Fixture file name contains facet id from: https://github.com/cursorless-dev/cursorless/issues/2010
* Scope is derived from facet id
* Fixture file only needs example code and the range representations described in this issue

example
`typescript/ternary.condition.txt`
```
;;!! true ? 0 : 1
;;!  ^^^^
```

**Comment by AndreasArvidsson** [comment]
Me and @josharian just had discussion about my post above. We are both in agreement that third and more verbose format is the way to go. We came up with a few thoughts.
* Only one range per header. If we need multiple content ranges we would have multiple content headers. This means that there is never range interleaving and there is no need to denote the start or end of a range. 
* The only part that's actually gonna be parsed/read from these files are the code at the top and to the delimiter we could have a custom format for the rest of the file if we like. If we in the future need a parsable format we could switch to something like [txtars](https://github.com/canopyclimate/txtars)
